Fixed #33278 -- Improved error for connection/query attempts against disallowed databases in tests.

SwastikTripathi opened the pull request:
[Ticket #33278](https://code.djangoproject.com/ticket/33278)
It should be clear at once if a database fails. Currently Database Failures raise AssertionError, and we have to trace the error.
A customized error type should make it a lot easier to use django.

felixxm commented on django/test/testcases.py:
```diff
@@ -145,7 +145,7 @@ def __init__(self, wrapped, message):
 
     def __call__(self):
-        raise AssertionError(self.message)
+        raise DatabaseFailureError(self.message)
```
`DatabaseFailureError` is not declared, it should be a subclass of `AssertionError`.

SwastikTripathi replied:
working on it :)

SwastikTripathi replied:
should I declare a subclass of `AssertionError` in `django/core/exceptions.py`?

felixxm reviewed (commented):
@SwastikTripathi Thanks for this patch :+1: Welcome aboard :boat: 

I renamed `DatabaseFailureError` to `DatabaseOperationForbidden` and switched `AssertionError` to the `DatabaseOperationForbidden` in tests.

SwastikTripathi commented on django/core/exceptions.py:
```diff
@@ -215,3 +215,7 @@ class EmptyResultSet(Exception):
     pass
+
+class DatabaseFailureError(AssertionError):
+    """Database blocked"""
```
I see. the error doesn't need to be accessible everywhere.

SwastikTripathi commented:
> @SwastikTripathi Thanks for this patch 👍 Welcome aboard ⛵
> 
> I renamed `DatabaseFailureError` to `DatabaseOperationForbidden` and switched `AssertionError` to the `DatabaseOperationForbidden` in tests.

`DatabaseOperationForbidden` is more specific. I agree it will be easier to understand
